nixos/dconf: support generating from attrs

linsui opened the pull request:
###### Description of changes

This implementation is partly based on https://github.com/NixOS/nixpkgs/pull/189099 and home-manager's dconf implementation and addresses some related reviews. A main difference between this and https://github.com/NixOS/nixpkgs/pull/189099 is that this implementation is completely backward-compatible and has lockdown support.

Example:

```nix
with lib.gvariant;
{
  enable = true;
  profiles.user.databases = [
    {
      settings = {
        "com/raggesilver/BlackBox" = {
          opacity = mkUint32 100;
          theme-dark = "Tommorow Night";
          scrollback-lines = mkUint32 10000;
        };
      };
      locks = [
        "/com/raggesilver/BlackBox/theme-dark"
      ];
    }
  ];
}
```

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2319

jtojnar commented:
Thanks. Will try to review this tomorrow or on Monday.

jtojnar commented on nixos/modules/programs/dconf.nix:
```diff
@@ -1,55 +1,174 @@
+                        description = lib.mdDoc "A list of dconf keyfiles.";
+                      };
+                      settings = lib.mkOption {
+                        type = lib.types.attrs;
```
Why `attrs` instead of just `attrsOf gvariant.value` like [home-manager has](https://nix-community.github.io/home-manager/options.html#opt-dconf.settings)? It would also be close to what comes from `dconf dump`.

linsui replied:
HM uses the path as the key but in this module the key is get from recursive attrs instead.

jtojnar replied:
Right. I am asking why was this decision made. Is it just to make locks implementation cleaner? Or are there some benefits for users?

linsui replied:
This design is inherited from https://github.com/NixOS/nixpkgs/pull/189099 and the locks implementation is based on this. I feel that this design can help organize the settings in a Nix style and the user can still choose to use the HM style.

> Is it just to make locks implementation cleaner?

Yes, looks like it's easier to implement the lockSubpaths in this design.

jtojnar replied:
Actually, one downside of nested attributes is that it no longer allows having both “files” and “directories“ of the same name, which `dconf` apparently supports. Not sure if any real application does this but if we want to allow creating arbitrary database, we probably cannot use this format.

linsui replied:
I'm not sure if I understand you correctly. Here is an example:

```
$ dconf dump /org/gnome/settings-daemon/plugins/
...
[media-keys]
...
custom-keybindings=['/org/gnome/settings-daemon/plugins/media-keys/custom-keybindings/custom0/']
...

[media-keys/custom-keybindings/custom0]
...
```

In this case we can't use

```nix
org.gnome.settings-daemon.plugins.media-keys.custom-keybindings = [ "/org/gnome/settings-daemon/plugins/media-keys/custom-keybindings/custom0/" ];
media-keys.custom-keybindings.custom0 = { ... };
```

but we can use

```nix
org.gnome.settings-daemon.plugins.media-keys.custom-keybindings = [ "/org/gnome/settings-daemon/plugins/media-keys/custom-keybindings/custom0/" ];
media-keys."custom-keybindings/custom0" = { ... };
```

instead. And if you want you can always use the `"path/to/the/directory"` format.

linsui replied:
The `subpathLocks` doesn't work with the slash format but this is expected.

jtojnar replied:
Well, currently, the option docs do not really mention slash format at all. But if we admit it, and it appears to work just as well as the nested format, as a user, I would expect it to work with locks as well.

But I am starting to think that maybe we should just use the slash format everywhere and use `lib.splitString` in the lock generator. I do not really see much of a benefit of the nested format for users and the slash format more closely resembles the keyfile contents so it will hopefully be less confusing.

linsui replied:
I can add support for that. Maybe it's easier to use something like `startswith` to generate locks.

jtojnar replied:
> The slash format is native to dconf keyfiles but the nested format is native to nix.

Yes, NixOS modules use nested attribute sets for their options but dconf is not NixOS. When modelling a domain, we should stick as closely as possible to it.

> With nested format, it's easier to take advantage of nix features without dealing with key strings. The lockSubpaths is an example.

I think we should leave `lockSubpaths` for a future PR. I do not like modules mixing custom NixOS-specific options with upstream ones and it could be easily replicated with something like:

```
rec {
  settings = {
    "foo/bar/baz" = {
      k1 = true;
    };
    "foo/bar/qux".k2 = true;
    "foo/miz/ter".k3 = "str";
  };
  locks =
    let
      lockedSubpaths = [
        "foo/bar"
      ];
      shouldBeLocked = path: lib.any (lockedSubpath: lib.hasPrefix "${lockedSubpath}/"  "${path}/") lockedSubpaths;
    in
    lib.concatMapAttrs (dirPath: keys: lib.concatMapAttrs (key: val: let path = "${dirPath}/${key}"; in { ${path} = shouldBeLocked path; }) keys) settings;
}
```

(In most cases, `shouldBeLocked` will just be `lib.const true` anyway.)

jtojnar replied:
Also reading https://discourse.nixos.org/t/write-key-value-using-lib-hm-gvariant-for-home-manager/31234, I realized we might eventually want `mkValue` to turn nix attrsets into GVariant dictionaries with string keys. If we allow nested keys, it would not be possible.

jtojnar commented on nixos/modules/programs/dconf.nix:
```diff
@@ -1,55 +1,174 @@
+  compileDconfDb = dir: pkgs.runCommand "dconf-db"
+    {
+      nativeBuildInputs = [ pkgs.dconf ];
+    } "dconf compile $out ${dir}";
```
Looks like this does not compile ibus database for some reason:

https://github.com/NixOS/nixpkgs/blob/486d2ce7d68af106bb5eccc46bc9f1fd4c7294b4/nixos/modules/i18n/input-method/ibus.nix#L66

jtojnar commented on nixos/modules/programs/dconf.nix:
```diff
@@ -1,55 +1,174 @@
+  # Generate Dconf profile
+  mkDconfProfile = name: value: pkgs.writeTextDir "etc/dconf/profile/${name}" (
+    if lib.isDerivation value || lib.isPath value then
+      lib.readFile value
```
We do not really allow import-from-derivation in Nixpkgs since it freezes Nix evaluation until the read file is built. Use `runCommand` with `mkdir` + `cp` instead of `writeTextDir` for this branch.

linsui replied:
Done. Why should we use cp instead of ln?

jtojnar replied:
I do not think it makes much of a difference. Mainly, when the link target is a derivation, it will end up in the output closure. This might be annoying when the package is large and nothing other than the dconf stuff is used from it but I do not think this will happen in real use cases.

jtojnar commented on nixos/modules/programs/dconf.nix:
```diff
@@ -1,55 +1,174 @@
+                        default = false;
+                        description = lib.mdDoc ''
+                          Similiar to locks but the whole subpath is locked. This
+                          only works for the dconf settings in databases.*.settings.
```
I wonder if we should even bother with this then. Users can pretty easily replicate this with `locks`.

Maybe once we have the `gsettings` module, we could have a `lockfiles` option and then have the gsettings module generate locks based on all existing keys.

linsui replied:
Could you elaborate this? I'd like to have this so that I can use `lockSubpaths = true` to lock all my settings.

jtojnar replied:
Ah. Yeah, that use case makes sense. I was only considering it in isolation.

jtojnar commented on lib/default.nix:
Might be cleaner to split it into multiple commits:

- GVariant functions
- dconf module refactoring
- dconf settings support
- dconf locks support
- gdm switch to settings

linsui replied:
Will do. For now I thought it's clearer for reviewers to just add modifications as commits?

jtojnar replied:
It can be but the downside is that it will need to be rebased before merging.
So it is usually fine to just amend the relevant commit and force push – unless you rebase onto a different base, GitHub will show you a clean diff of the changes.

linsui replied:
Good point! Will do.

jtojnar commented on nixos/modules/programs/dconf.nix:
```diff
@@ -1,55 +1,174 @@
+    } "dconf compile $out ${dir}";
+
+  # Recursively lock values in settings
+  recuriseLock = with lib; settings: locks: (mapAttrsRecursive
+    (path: value: optionalAttrs value
+      (mapAttrsRecursiveCond (value: !lib.gvariant.isGVariant value)
+        (path: value: true)
+        (foldl'
+          (root: name: if hasAttr "${name}" root then root."${name}" else
+          throw "${concatStringsSep "." (tail path)} doesn't exist in settings.")
+          { v = settings; }
+          path)))
+    { v = locks; }).v;
```
It is very difficult to see which attributes belong to which function. Could you please indent it so it is clearer, for example:

```suggestion
  recuriseLock =
    settings:
    locks:

    (lib.mapAttrsRecursive
      (path:
      value:

      lib.optionalAttrs
        value
        (lib.mapAttrsRecursiveCond
          (value': !lib.gvariant.isGVariant value')
          (path': value': true)
          (foldl'
            (root: name: root."${name}" or throw "${lib.concatStringsSep "." (lib.tail path)} doesn't exist in settings.")
            { v = settings; }
            path
          )
        )
      )
      { v = locks; }).v;
```

Or consider splitting the function somehow.

linsui commented on nixos/modules/programs/dconf.nix:
```diff
@@ -1,55 +1,174 @@
   };
 
-  ###### implementation
+  config = lib.mkIf cfg.enable {
```
Should we enable dconf explicitly when setting profiles instead?

linsui replied:
Reverted.

linsui commented on nixos/modules/programs/dconf.nix:
```diff
@@ -1,55 +1,174 @@
+        description = lib.mdDoc "Attrset of dconf profiles.";
+      };
+
+      defaultProfile = lib.mkOption {
```
This is taken from https://github.com/NixOS/nixpkgs/pull/189099. You are the expert. ;) I can add this as a document instead.

linsui replied:
Removed.

linsui commented on lib/gvariant.nix:
```diff
@@ -0,0 +1,181 @@
+  mkValue = v:
+    if builtins.isBool v then
+      mkBoolean v
+    else if builtins.isInt v then
+      mkInt32 v
+    else if builtins.isFloat v then
+      mkDouble v
```
I removed this. What's the GSettings module?

jtojnar replied:
In the future, GSettings module could use the dconf module to set settings and validate them against GSettings schemas. That way we would not need to care about which integer type is expected, we could just find out from the schema.

But that is orthogonal and removal makes sense even if `dconf` becomes only a low-level module.

linsui commented on lib/gvariant.nix:
```diff
@@ -0,0 +1,181 @@
+# This file is taken from https://github.com/nix-community/home-manager
```
Where should I put the test? Should I add `lib/tests/gvariant.nix`?

jtojnar replied:
If you copy https://github.com/nix-community/home-manager/blob/5db22bce05c776057fdb289da17f6c12049c4624/tests/lib/types/gvariant-merge.nix (replacing `home.file` + `nmt.scripts` by direct assertions, it could probably go to to `lib/tests/modules/gvariant.nix`.

Commit:
lib/gvariant: init

Commit:
nixos/dconf: refractor

remove `with lib;`
profiles option now accepts packages in addition to paths.
profiles option is no longer internal.
cfgDir definition has been inlined.
pulled GIO_EXTRA_MODULES inside mkif.
removed pointless comments with section headings.
defined profiles are now turned into package, allowing to simplify the db update logic.

Commit:
nixos/dconf: add locks support

Commit:
nixos/gdm: switch to dconf settings

linsui commented on nixos/modules/programs/dconf.nix:
```diff
@@ -1,55 +1,174 @@
+
+                  A key is searched from up to down and the first result takes the
+                  priorit. If a lock for a particular key is installed then the
+                  last locked key takes the priorit. This can be used to enforce
```
I'm not sure what you mean. What's your suggestion?

jtojnar replied:
I am not sure what this sentence means: “If a lock for a particular key is installed then the last locked key takes the priorit”

linsui replied:
I try to adapt this part in https://wiki.gnome.org/Projects/dconf/SystemAdministrators

> You can actually think of the search for a dconf key as starting in the lowest database and checking for locked keys all the way to the top level database, stopping if it finds one. If no locked keys are found then dconf searches the user's database and then back down the profile again for unlocked keys. Stated another way, locks installed into databases lower in the profile take precedence over locks installed in higher databases.

jtojnar replied:
Maybe something like “If a lock for a particular key is installed then the value from the last database in the profile where the key is locked will be used.” would be clearer?

linsui replied:
Thanks!

linsui commented on lib/gvariant.nix:
```diff
@@ -0,0 +1,181 @@
+  # Returns the GVariant value that most closely matches the given Nix
+  # value. If no GVariant value can be found then `null` is returned.
+
+  mkValue = v:
```
I added docs for every functions.

jtojnar replied:
Nice. Please also add it to https://github.com/NixOS/nixpkgs/blob/2ba28b9c95690e40187c99d6756e98ac58d678b2/doc/doc-support/default.nix#L6 so that it is included in the Nixpkgs manual (buildable with `nix-build pkgs/top-level/release.nix -A manual`).

linsui commented on nixos/modules/services/x11/display-managers/gdm.nix:
```diff
@@ -229,40 +229,15 @@ in
-    in pkgs.stdenv.mkDerivation {
-      name = "dconf-gdm-profile";
-      buildCommand = ''
-        # Check that the GDM profile starts with what we expect.
```
Do you mean that I should add the profile content into passthru then check it in the module?

jtojnar replied:
I meant adding the profile content into `passthru`:

```nix
passthru = {
  dconfDb = "${finalAttrs.finalPackage}/share/gdm/greeter-dconf-defaults";
  dconfProfile = ''
    user-db:user
    file-db:${finalAttrs.passthru.dconfDb}
  '';
};
```

checking it in `passthru.tests`, and using `gdm.dconfDb`  in the NixOS module.

It will require switching the package to `finalAttrs` first (see e.g. https://github.com/NixOS/nixpkgs/pull/237140).

roberth commented on lib/generators.nix:
```diff
@@ -175,6 +175,41 @@ rec {
+     Type:
+       flattenAttrsCond :: (AttrSet -> Bool) -> ([String] -> a -> AttrSet) -> AttrSet -> AttrSet
+  */
+  flattenAttrsCond =
```
There's another one as well. I've turned the list into an issue for further discussion.
 - https://github.com/NixOS/nixpkgs/issues/237776

I think it'd be kinda ok to keep the function in a local let binding with a comment not to expose it until https://github.com/NixOS/nixpkgs/issues/237776 is resolved, although otoh it's not clear when the added tech debt will be resolved.

linsui replied:
I put it into let-in.

jtojnar commented on pkgs/desktops/gnome/core/gdm/default.nix:
```diff
@@ -151,13 +159,7 @@ stdenv.mkDerivation rec {
-  # at install time but Meson does not support this
-  # so we need to convince it to install all files to a temporary
-  # location using DESTDIR and then move it to proper one in postInstall.
-  DESTDIR = "${placeholder "out"}/dest";
```
Why move this?

linsui replied:
I thought it's easier to put this into let in instead of write finalAttrs everywhere.

jtojnar replied:
It does not use `finalAttrs` though?

linsui replied:
I need to use `${finalAttrs.DESTDIR}`. Should I use `$DESTDIR` instead?

jtojnar replied:
Ah, the `postInstall`. It should not really be interpolating the value in the first place, picking it up from the environment `$DESTDIR` should be sufficient).

linsui commented on nixos/modules/programs/dconf.nix:
```diff
@@ -1,55 +1,58 @@
+        name = "dconf-system-config";
+        paths = map (x: "${x}/etc/dconf") cfg.packages;
+        nativeBuildInputs = [ (lib.getBin pkgs.dconf) ];
+        postBuild = "dconf update $out/db";
```
It's $out/db since it's the source of etc.dconf.

jtojnar replied:
You are right, I misread `paths` as `pathsToLink`.

linsui commented on nixos/modules/programs/dconf.nix:
```diff
@@ -3,12 +3,94 @@
+  compileDconfDb = dir: pkgs.runCommand "dconf-db"
+    {
+      nativeBuildInputs = [ (lib.getBin pkgs.dconf) ];
+    } "dconf compile $out ${dir}";
```
The `dconf update` is used for packages with profiles and dconf keyfiles. The database is generated into the same directory in `/etc/dconf` as system-db.

The `dconf compile` is used to generated a database ad add it to a profile as a file-db. Only the profile is added to `/etc/dconf`.

linsui commented on lib/gvariant.nix:
```diff
@@ -0,0 +1,256 @@
+     Type:
+       isGVariant :: [Any] -> AttrSet
+  */
+  mkDictionaryEntry = elems:
```
Where are the names used?

jtojnar replied:
They would be used when calling the function. We would require user to use `mkDictionaryEntry { name = "foo"; value = "bar"; }` rather than `mkDictionaryEntry [ "foo" "bar" ]`.

linsui commented on nixos/modules/programs/dconf.nix:
```diff
@@ -3,12 +3,94 @@
+      databases = lib.mkOption {
+        type = with lib.types; listOf (oneOf [
+          path
+          package
```
The path is writen to the profile as a file-db. Do I need to make sure it's a file? I thought the user should ensure that it's a valid dconf database.

jtojnar replied:
Yeah, ideally, the API should prevent errors as soon as possible, when practical. For example, if user passes `dbPackage` that really contains a directory with `db.conf`, it would fail. Though, I am not sure if it is possible to distinguish that. Maybe it is actually not practical :woman_shrugging:

linsui replied:
Maybe we can write a key to the database.

jtojnar replied:
Not sure what you mean by that.

linsui replied:
My fault. `dconf write` can't write to a specfic database file. I thought we can only check if it's a file. Should I add that?

jtojnar replied:
Right, `dconf write` only touches `user-db`. I am still not sure how writing is relevant to linking databases from profile.

File check would be nice but it is probably not possible to do it at module system type level (without IFD).

Maybe replacing `pkgs.writeTextDir "etc/dconf/profile/${name}"` with `writeTextFile` and checking in its `checkPhase` would work, though.

linsui replied:
How about checking and copying in runCommand?

linsui replied:
Now it's checked in runCommand to be a valid dconf database.

jtojnar replied:
Thanks. That should work.

linsui commented on nixos/modules/programs/dconf.nix:
```diff
@@ -1,55 +1,58 @@
 { config, lib, pkgs, ... }:
```
It's mostly removing the `with lib;`.

jtojnar commented on lib/gvariant.nix:
```diff
@@ -0,0 +1,262 @@
+  /* Returns the GVariant dictionary entry from the given Nix list.
+
+     Type:
+       mkDictionaryEntry :: { name :: Any; value :: Any } -> gvariant
```
The curried form is also an option but in that case the type annotation should also reflect that (`Any -> Any -> gvariant`).

jtojnar commented on nixos/modules/programs/dconf.nix:
```diff
@@ -1,55 +1,182 @@
-          ''
-        ) cfg.profiles
+
+  # Compile keyfiles to dconf DB
```
Maybe include a mention of something like “expects `/etc/dconf/db/*.d` directory“ so that it is clearer to visualise what is happening when wrapping ones head around the code.

jtojnar commented on nixos/modules/programs/dconf.nix:
```diff
@@ -1,55 +1,182 @@
-    '';
+      { v = locks; }).v;
+
+  # Generate dconf DB from attrset and keyfiles
```
Then it will be possible to say that `val` has the type of `dconfDatabase` submodule.

linsui replied:
Do you mean adding it to the comment?

linsui commented on nixos/modules/programs/dconf.nix:
```diff
@@ -3,12 +3,94 @@
+          (submodule {
+            options = {
+              keyfiles = lib.mkOption {
+                type = with lib.types; listOf (oneOf [ path package ]);
```
The keyfiles are checked that they can be compiled.

linsui commented:
@jtojnar Could you please take another look?

jtojnar commented on lib/gvariant.nix:
```diff
@@ -0,0 +1,266 @@
+  /* Returns the GVariant dictionary entry from the given key and value.
+
+     Type:
+       mkDictionaryEntry :: Any -> Any -> gvariant
```
```suggestion
       mkDictionaryEntry :: Any -> Any -> gvariant
       
     Example:
       # A dictionary describing an Epiphany’s search provider
       [
         (lib.gvariant.mkDictionaryEntry "url" (lib.gvariant.mkVariant "https://duckduckgo.com/?q=%s&t=epiphany"))
         (lib.gvariant.mkDictionaryEntry "bang" (lib.gvariant.mkVariant "!d"))
         (lib.gvariant.mkDictionaryEntry "name" (lib.gvariant.mkVariant "DuckDuckGo"))
       ]
```

jtojnar commented on lib/gvariant.nix:
```diff
@@ -0,0 +1,266 @@
+# A partial and basic implementation of GVariant formatted strings. See https://docs.gtk.org/glib/gvariant-format-strings.html
+# for detauls.
+#
+# Note, this API is not considered fully stable and it might therefore
+# change in backwards incompatible ways without prior notice.
```
This should probably be included in the docs as well.

jtojnar replied:
And maybe we should express it also in the attribute path, see also `pkgs._experimental-update-script-combinators`. Not completely sure about this.

linsui replied:
Where should I add it? Is it enough to convert it to a `/**/` style comment?

jtojnar replied:
Looks like nixdoc currently does not support module docs. Will try to implement it.

linsui replied:
Do we need to wait for that?

jtojnar commented on nixos/modules/programs/dconf.nix:
```diff
@@ -1,55 +1,245 @@
+
+    echo "file-db:${file}" > profile
+    DCONF_PROFILE=$(pwd)/profile dconf dump / > dump 2> error
+    if [[ ! -s output ]] && [[ -s error ]]; then
```
Empty databases are apparently possible

```ShellSession
$ cd "$(mktemp -d)"
$ mkdir -p foo.d
$ touch foo.d/foo
$ dconf compile foo foo.d
$ echo file-db:foo > profile
$ DCONF_PROFILE=$PWD/profile dconf dump /
```

Not sure if this edge case is worth handling, though.

linsui replied:
Yes, the database can be empty but looks like it makes no sense to add an empty database.

linsui replied:
And I also check that there is no error so this should have been covered.

jtojnar replied:
You are right. That should work.

jtojnar commented on nixos/modules/programs/dconf.nix:
```diff
@@ -1,55 +1,245 @@
+
+    echo "file-db:${file}" > profile
+    DCONF_PROFILE=$(pwd)/profile dconf dump / > dump 2> error
+    if [[ ! -s output ]] && [[ -s error ]]; then
```
Also you are writing the stdout to `dump` file but checking `output`.

jtojnar commented on nixos/modules/programs/dconf.nix:
Please describe other changes in the `nixos/dconf: refractor` commit message as well, the commit mixes so many changes it is difficult to decode them. Also typo ref**r**actor.

From what I see:

- `profiles` option now accepts packages in addition to paths.
- `profiles` option is no longer internal.
- `cfgdir` definition has been inlined.
- pulled `GIO_EXTRA_MODULES` inside `mkif`.
- removed pointless comments with section headings.
- defined profiles are now turned into package, allowing to simplify the db update logic.

jtojnar commented on nixos/modules/programs/dconf.nix:
```diff
@@ -3,12 +3,158 @@
+      nativeBuildInputs = [ (lib.getBin pkgs.dconf) ];
+    } "dconf compile $out ${dir}";
+
+  # Recursively lock values in settings
```
Looks like this should go to the lock commit?

jtojnar commented on nixos/modules/programs/dconf.nix:
```diff
@@ -3,12 +3,158 @@
+    fi
+
+    dconf compile db ${dir} || (
+      echo "The Dconf keyfiles are invalid: ${dir}"
```
dconf should be lowercase.

jtojnar reviewed (commented):
Sorry for the wait, here is another batch of reviews. It has improved but I am still not very happy about the nesting.

linsui commented:
> Also reading https://discourse.nixos.org/t/write-key-value-using-lib-hm-gvariant-for-home-manager/31234, I realized we might eventually want mkValue to turn Nix attrsets into GVariant dictionaries with string keys. If we allow nested keys, it would not be possible.

I thought we need to convert the attrs to a GVariant dictionaries first anyway.

linsui commented:
> Yes, NixOS modules use nested attribute sets for their options but dconf is not NixOS. When modelling a domain, we should stick as closely as possible to it.

But I thought the whole point of `toINI` is expressing ini files with native nix stype attrs. It's reasonable to me to convert the dconf path as a nested attrs.

> I think we should leave lockSubpaths for a future PR

Yes, OK to me.

linsui commented on lib/gvariant.nix:
```diff
@@ -0,0 +1,266 @@
+    __toString = self: "@${self.type} ${toString self.value}"; # https://docs.gtk.org/glib/gvariant-text.html
+  };
+
+  type = {
```
I thought this is an internal implementation detail a doc comment is enough.

jtojnar replied:
It is exported and being used in stuff like `mkEmptyArray`.

linsui replied:
Ah, right. Let me check if I can document it.

bam80 replied:
I'm not sure was it documented after all?
I don't see any evidence of touching documentation.

jtojnar commented:
> > Also reading [discourse.nixos.org/t/write-key-value-using-lib-hm-gvariant-for-home-manager/31234](https://discourse.nixos.org/t/write-key-value-using-lib-hm-gvariant-for-home-manager/31234), I realized we might eventually want mkValue to turn Nix attrsets into GVariant dictionaries with string keys. If we allow nested keys, it would not be possible.
> 
> I thought we need to convert the attrs to a GVariant dictionaries first anyway.

Right but we might still implement it with something like the following:

```patch
@@ -68,6 +69,11 @@ rec {
       else mkArray v
     else if isGVariant v then
       v
+    else if builtins.isAttrs v then
+      lib.throwIf
+        (v == {})
+        "Please create an empty dictionary with something like `lib.gvariant.mkEmptyArray (lib.gvariant.type.dictionaryEntryOf lib.gvariant.type.string lib.gvariant.type.int32)`"
+        (lib.mapAttrsToList mkDictionaryEntry v)
     else
       throw "The GVariant type of ${v} can't be inferred.";
 
```

But I have not considered all the downsides yet so it is probably best left for future PRs as well.

---

> > Yes, NixOS modules use nested attribute sets for their options but dconf is not NixOS. When modelling a domain, we should stick as closely as possible to it.
> 
> But I thought the whole point of `toINI` is expressing ini files with native nix stype attrs. It's reasonable to me to convert the dconf path as a nested attrs.

The use of `toINI` is implementation detail IMO – the point is modelling dconf database contents in Nix. With that in mind, I think automatically converting leaf values makes sense for convenience, just like we do with [RFC 42 settings](https://github.com/NixOS/rfcs/pull/42) . But I am still not convinced replicating the hierarchy is needed.

In INI-based settings, you often have two levels: sections and keys. dconf appears hierarchical but in practice, people tend to think in two levels as well: GSettings schemas (analogous to INI sections, as they are used as such in keyfiles) and keys. This is also evidenced in the way you write the examples:

```nix
test.not.locked = mkInt32 1;
test.is.locked = "locked";
```

not

```nix
test = {
  not.locked = mkInt32 1;
  is.locked = "locked";
};
```

(Though I admit nesting would feel natural in some use cases, e.g. with relocatable schemas. But those are comparatively rare so I am still not convinced the added possibilities for silent misconfiguration is worth it.)

(Actually, dconf describes itself as a key database, which is also shown in commands `dconf read /org/gnome/desktop/wm/keybindings/move-to-center`. So maybe we should go even further and have the settings be single level, and leave the nesting for future GSettings schema-based higher-level module.)

See also the `environment.etc` NixOS option for another instance of this pattern.

I am not saying I will never be convinced the other way but for now I think the paths as strings close fewer doors. It can always be extended with nesting-attributes paths support or `mkValue` for dictionaries later, while the other way is not possible.

linsui committed:
nixos/dconf: add settings support

linsui commented:
I removed all features related to the nested format.

jtojnar reviewed (approved):
Looks perfect to me now. Thanks for bearing with me.

jtojnar commented:
Looks like the tests do not actually run.

Also, I am thinking of getting rid of `mkEmptyArray` since it needs to be replaced `mkArray` when one decides to to add elements. Something like the following.


```patch
diff --git a/lib/gvariant.nix b/lib/gvariant.nix
index 3142ffc5f149..06bd3ef6a297 100644
--- a/lib/gvariant.nix
+++ b/lib/gvariant.nix
@@ -66,7 +66,11 @@ rec {
     else if builtins.isString v then
       mkString v
     else if builtins.isList v then
-      mkArray v
+      let
+        vs = map mkValue v;
+        head = lib.throwIf (v == [ ]) "Attempting to create an empty array but mkValue does not allow specifying type. Please create empty array with mkArray and pass the type explicitly." (builtins.head vs);
+      in
+      mkArray head.type vs
     else if isGVariant v then
       v
     else
@@ -75,15 +79,15 @@ rec {
   /* Returns the GVariant array from the given type of the elements and a Nix list.
 
      Type:
-       mkArray :: [Any] -> gvariant
+       mkArray :: gvariant.type -> [Any] -> gvariant
 
      Example:
        # Creating a string array
-       lib.gvariant.mkArray [ "a" "b" "c" ]
+       lib.gvariant.mkArray  [ "a" "b" "c" ]
   */
-  mkArray = elems:
+  mkArray = elemType: elems:
     let
-      vs = map mkValue (lib.throwIf (elems == [ ]) "Please create empty array with mkEmptyArray." elems);
+      vs = map mkValue elems;
       elemType = lib.throwIfNot (lib.all (t: (head vs).type == t) (map (v: v.type) vs))
         "Elements in a list should have same type."
         (head vs).type;
@@ -93,19 +97,6 @@ rec {
         "@${self.type} [${concatMapStringsSep "," toString self.value}]";
     };
 
-  /* Returns the GVariant array from the given empty Nix list.
-
-     Type:
-       mkEmptyArray :: gvariant.type -> gvariant
-
-     Example:
-       # Creating an empty string array
-       lib.gvariant.mkEmptyArray (lib.gvariant.type.string)
-  */
-  mkEmptyArray = elemType: mkPrimitive (type.arrayOf elemType) [ ] // {
-    __toString = self: "@${self.type} []";
-  };
-
 
   /* Returns the GVariant variant from the given Nix value. Variants are containers
      of different GVariant type.
diff --git a/lib/tests/modules/gvariant.nix b/lib/tests/modules/gvariant.nix
index a792ebf85b74..d2bbbbcc88b2 100644
--- a/lib/tests/modules/gvariant.nix
+++ b/lib/tests/modules/gvariant.nix
@@ -31,12 +31,12 @@ in {
         { uint64 = mkUint64 42; }
 
         { array1 = [ "one" ]; }
-        { array1 = mkArray [ "two" ]; }
-        { array2 = mkArray [ (mkInt32 1) ]; }
-        { array2 = mkArray [ (nkUint32 2) ]; }
+        { array1 = mkArray type.string [ "two" ]; }
+        { array2 = mkArray type.int32 [ (mkInt32 1) ]; }
+        { array2 = mkArray type.uint32 [ (nkUint32 2) ]; }
 
         { emptyArray1 = [ ]; }
-        { emptyArray2 = mkEmptyArray type.uint32; }
+        { emptyArray2 = mkArray type.uint32 []; }
 
         { string = "foo"; }
         { string = "foo"; }
```

linsui commented:
So when the type can be infered the user don't need to use mkArray and mkArray is only used for empty list?

jtojnar commented:
Yes.

On Tue, 12 Sept 2023, 04:19 linsui, ***@***.***> wrote:

> So when the type can be infered the user don't need to use mkArray and
> mkArray is only used for empty list?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/234615#issuecomment-1714865174>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFMEY4MAKUS27TXX4MDRNTXZ7BDZANCNFSM6AAAAAAYR2S26Y>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>

linsui commented:
Then I thought the situation doesn't change. Currently if the user want to add elements to an empty array they need to remove the mkEmptyArray and type. With your proposal, the user needs to remove the mkArray and type.

jtojnar commented:
They do not need to, the explicit constructor is fine to use.

linsui commented:
How about making mkArray only require the type when creating an empty array?

linsui commented:
> Looks like the tests do not actually run.

How can I fix this?

jtojnar commented:
> How about making mkArray only require the type when creating an empty array?

I think having functions that take different number of arguments depending what those arguments are can be confusing (e.g. it is not possible to express it using type annotations).



> > Looks like the tests do not actually run.
> 
> How can I fix this?

One way would be adding assertions to https://github.com/NixOS/nixpkgs/blob/fff294ce97764ee8ec20aeb24b20fff114aebbd2/lib/tests/modules.sh

The other would be creating file like https://github.com/NixOS/nixpkgs/blob/fff294ce97764ee8ec20aeb24b20fff114aebbd2/lib/tests/misc.nix and then adding it to https://github.com/NixOS/nixpkgs/blob/55ec5ae7d6c3f7866a0696a6ccfb66a1665b3d72/lib/tests/check-eval.nix#L3

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/configuration-of-gnome-extensions/33337/2

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/need-help-for-nixos-gnome-scaling-settings/24590/12
